### ISSUE TITLE
BigQueryDataset.spec.access is not working properly.

### ISSUE THREAD

**Issue by rnaveiras** [GoogleCloudPlatform/k8s-config-connector#291]
**Describe the bug**

When we tried to use `BigQueryDataset.spec.access` to grant access to the dataset the config connector gets stuck in the reconciliation loop with a panic and it never updates the resource in BigQuery.

**ConfigConnector Version**
Version 1.24.0

**To Reproduce**
Use the following YAML snippet produces the error shown above.

```yaml
---
apiVersion: bigquery.cnrm.cloud.google.com/v1beta1
kind: BigQueryDataset
metadata:
  annotations:
    cnrm.cloud.google.com/delete-contents-on-destroy: "false"
  name: foo
spec:
  access:
  - role: OWNER
    specialGroup: projectOwners
  - groupByEmail: team@example.com
    role: READER
  - role: WRITER
    userByEmail: service-account-email@project-id.iam.gserviceaccount.com
  - role: WRITER
     userByEmail: other-service-account-email@project-id.iam.gserviceaccount.com
  friendlyName: foo
  location: EU
```

Log error: 

```json
{
  "severity": "error",
  "logger": "controller-runtime.controller",
  "msg": "Reconciler error",
  "controller": "bigquerydataset-controller",
  "request": "abacus-sandbox-staging/abacus",
  "error": "INTERNAL ERROR: panic: observed a panic: expected state's set to have value explicitly provided in spec\n\t/go/src/cnrm.googlesource.com/cnrm/pkg/execution/builtin.go:26 +0x57\n\tpanic(0x33ef6e0, 0xc007baee50)\n\t\t/usr/local/go/src/runtime/panic.go:975 +0x3e3\n\tcnrm.googlesource.com/cnrm/pkg/krmtotf.convertTFSetToKCCSet(0x318c260, 0xc0054d3d80, 0x318c260, 0xc00255c6a0, 0xc000ab2000, 0xc000e67cb0, 0x3a0280f, 0x6, 0x51849a, 0xc001360918)\n\t\t/go/src/cnrm.googlesource.com/cnrm/pkg/krmtotf/tftokrm.go:514 +0xa82\n\tcnrm.googlesource.com/cnrm/pkg/krmtotf.convertTFMapToKCCMap(0xc002ee6cf0, 0xc002c6d350, 0xc000f39590, 0xc000e67cb0, 0x0, 0x0, 0xc002c6ca20, 0x0)\n\t\t/go/src/cnrm.googlesource.com/cnrm/pkg/krmtotf/tftokrm.go:338 +0xa8c\n\tcnrm.googlesource.com/cnrm/pkg/krmtotf.ConvertTFObjToKCCObj(0xc002ee6cf0, 0xc002c6d350, 0xc000f39590, 0xc000e67cb0, 0x0, 0x0, 0xc002c6ca20, 0xc000cf0a20)\n\t\t/go/src/cnrm.googlesource.com/cnrm/pkg/krmtotf/tftokrm.go:230 +0x7a\n\tcnrm.googlesource.com/cnrm/pkg/krmtotf.GetSpecAndStatusFromState(0xc000e67b80, 0xc002c2dea0, 0x0, 0x0)\n\t\t/go/src/cnrm.googlesource.com/cnrm/pkg/krmtotf/tftokrm.go:24 +0xcd\n\tcnrm.googlesource.com/cnrm/pkg/krmtotf.resolveUnmanagedFields(0xc004651110, 0xc000e67b80, 0xc002c2dea0, 0x0, 0x0, 0xc000cf0698)\n\t\t/go/src/cnrm.googlesource.com/cnrm/pkg/krmtotf/gcpmanagedfields.go:287 +0xb6\n\tcnrm.googlesource.com/cnrm/pkg/krmtotf.krmResourceToTFResourceConfig(0xc000e67b80, 0xc002c2dea0, 0x41bf860, 0xc0029b0820, 0xc0006ebc40, 0x1, 0x0, 0xc002c2dea0, 0x0, 0x0)\n\t\t/go/src/cnrm.googlesource.com/cnrm/pkg/krmtotf/krmtotf.go:57 +0x17d\n\tcnrm.googlesource.com/cnrm/pkg/krmtotf.KRMResourceToTFResourceConfigWithGCPManagedFields(...)\n\t\t/go/src/cnrm.googlesource.com/cnrm/pkg/krmtotf/krmtotf.go:46\n\tcnrm.googlesource.com/cnrm/pkg/controller/tf.(*Reconciler).sync(0xc0029b0820, 0x41a3ca0, 0xc0001901d8, 0xc000e67b80, 0x41e8300, 0x0, 0x0)\n\t\t/go/src/cnrm.googlesource.com/cnrm/pkg/controller/tf/controller.go:234 +0x546\n\tcnrm.googlesource.com/cnrm/pkg/controller/tf.(*Reconciler).Reconcile(0xc0029b0820, 0xc000f615e0, 0x16, 0xc00320ead8, 0x6, 0x0, 0x0, 0x0, 0x0)\n\t\t/go/src/cnrm.googlesource.com/cnrm/pkg/controller/tf/controller.go:147 +0xbf3\n\tcnrm.googlesource.com/cnrm/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc002a500c0, 0x3502fe0, 0xc0055d14a0, 0x3924900)\n\t\t/go/src/cnrm.googlesource.com/cnrm/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:256 +0x161\n\tcnrm.googlesource.com/cnrm/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc002a500c0, 0x203000)\n\t\t/go/src/cnrm.googlesource.com/cnrm/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:232 +0xae\n\tcnrm.googlesource.com/cnrm/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker(0xc002a500c0)\n\t\t/go/src/cnrm.googlesource.com/cnrm/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:211 +0x2b\n\tcnrm.googlesource.com/cnrm/vendor/k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc001c45070)\n\t\t/go/src/cnrm.googlesource.com/cnrm/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x5f\n\tcnrm.googlesource.com/cnrm/vendor/k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc001c45070, 0x41161a0, 0xc0026b6840, 0xc000cd0101, 0xc00074e960)\n\t\t/go/src/cnrm.googlesource.com/cnrm/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0xa3\n\tcnrm.googlesource.com/cnrm/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc001c45070, 0x3b9aca00, 0x0, 0x2020202020202001, 0xc00074e960)\n\t\t/go/src/cnrm.googlesource.com/cnrm/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0xe2\n\tcnrm.googlesource.com/cnrm/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc001c45070, 0x3b9aca00, 0xc00074e960)\n\t\t/go/src/cnrm.googlesource.com/cnrm/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90 +0x4d\n\tcreated by cnrm.googlesource.com/cnrm/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\t\t/go/src/cnrm.googlesource.com/cnrm/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:193 +0x305\n\t"
}
```

https://cloud.google.com/config-connector/docs/reference/resource-docs/bigquery/bigquerydataset

**Comment by caieo** [comment]
Hi @rnaveiras , thank you for reporting this bug. I was unable to reproduce it following the YAML example you provided, but did encounter it accidentally. In the last entry of `spec.access`, I added the same service account e-mail which caused that error to appear (and once I replaced it with a different service account, the error disappeared). My hunch is that the expected set is removing the duplicated value and therefore causing a panic between the expected and explicit set. Can you check if the service accounts you use for the `WRITER` access roles are different?

**Comment by xiaobaitusi** [comment]
Hey @rnaveiras, does caie@'s last comment help with solving the issue? Let us know if you are still blocked.

**Comment by tkinz27** [comment]
I am having the same problem with a simpler manifest.

```yaml
---
apiVersion: bigquery.cnrm.cloud.google.com/v1beta1
kind: BigQueryDataset
metadata:
  annotations:
    cnrm.cloud.google.com/project-id: "projectid"
    cnrm.cloud.google.com/delete-contents-on-destroy: "true"
  name: datapoc
spec:
  friendlyName: data-poc
  location: us-west1
  description: An example of loading data into big query
  defaultPartitionExpirationMs: 604800000
  access:
    - role: roles/bigquery.dataEditor
      userByEmail: bq-transfer-service@projectid.iam.gserviceaccount.com
```

```
{"severity":"error","logger":"controller-runtime.controller","msg":"Reconciler error","controller":"bigquerydataset-controller","request":"data-poc/datapoc","error":"INTERNAL ERROR: panic: observed a panic: expected state's set to have value explicitly provided in spec\n\t/go/src/cnrm.googlesource.com/cnrm/pkg/execution/builtin.go:26 +0x57\n\tpanic(0x33f67e0, 0xc003f45960)\n\t\t/usr/local/go/src/runtime/panic.go:975 +0x3e3\n\tcnrm.googlesource.com/cnrm/pkg/krmtotf.convertTFSetToKCCSet(0x3192820, 0xc003f15fe0, 0x3192820, 0xc003dcb200, 0xc000bf1a40, 0xc003b251b8, 0x3a0a504, 0x6, 0x51849a, 0xc003ca2ff8)\n\t\t/go/src/cnrm.googlesource.com/cnrm/pkg/krmtotf/tftokrm.go:502 +0xa82\n\tcnrm.googlesource.com/cnrm/pkg/krmtotf.convertTFMapToKCCMap(0xc003f3e690, 0xc003dcd950, 0xc000be4600, 0xc003b251b8, 0x0, 0x0, 0xc003dcd0b0, 0x0)\n\t\t/go/src/cnrm.googlesource.com/cnrm/pkg/krmtotf/tftokrm.go:338 +0xa8c\n\tcnrm.googlesource.com/cnrm/pkg/krmtotf.ConvertTFObjToKCCObj(0xc003f3e690, 0xc003dcd950, 0xc000be4600, 0xc003b251b8, 0x0, 0x0, 0xc003dcd0b0, 0xc003dc9440)\n\t\t/go/src/cnrm.googlesource.com/cnrm/pkg/krmtotf/tftokrm.go:230 +0x7a\n\tcnrm.googlesource.com/cnrm/pkg/krmtotf.GetSpecAndStatusFromState(0xc003b25080, 0xc003dae540, 0x0, 0x0)\n\t\t/go/src/cnrm.googlesource.com/cnrm/pkg/krmtotf/tftokrm.go:24 +0xcd\n\tcnrm.googlesource.com/cnrm/pkg/krmtotf.resolveUnmanagedFields(0xc003f2ae40, 0xc003b25080, 0xc003dae540, 0x0, 0x0, 0x22ecb25c000)\n\t\t/go/src/cnrm.googlesource.com/cnrm/pkg/krmtotf/krmtotf.go:145 +0xb6\n\tcnrm.googlesource.com/cnrm/pkg/krmtotf.krmResourceToTFResourceConfig(0xc003b25080, 0xc003dae540, 0x41ca2c0, 0xc0021ddf40, 0xc00060c610, 0x1, 0x0, 0xc003dae540, 0x0, 0x0)\n\t\t/go/src/cnrm.googlesource.com/cnrm/pkg/krmtotf/krmtotf.go:57 +0x17d\n\tcnrm.googlesource.com/cnrm/pkg/krmtotf.KRMResourceToTFResourceConfigWithGCPManagedFields(...)\n\t\t/go/src/cnrm.googlesource.com/cnrm/pkg/krmtotf/krmtotf.go:46\n\tcnrm.googlesource.com/cnrm/pkg/controller/tf.(*Reconciler).sync(0xc0021ddf40, 0x41ae6c0, 0xc0001241e8, 0xc003b25080, 0x41f2f00, 0x0, 0x0)\n\t\t/go/src/cnrm.googlesource.com/cnrm/pkg/controller/tf/controller.go:229 +0x546\n\tcnrm.googlesource.com/cnrm/pkg/controller/tf.(*Reconciler).Reconcile(0xc0021ddf40, 0xc002756ec0, 0xb, 0xc002756ea0, 0xe, 0x0, 0x0, 0x0, 0x0)\n\t\t/go/src/cnrm.googlesource.com/cnrm/pkg/controller/tf/controller.go:142 +0xbf3\n\tcnrm.googlesource.com/cnrm/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc00263e600, 0x350a1a0, 0xc003e3b580, 0x4670ef3eb00)\n\t\t/go/src/cnrm.googlesource.com/cnrm/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:256 +0x161\n\tcnrm.googlesource.com/cnrm/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc00263e600, 0x4670ef3eb00)\n\t\t/go/src/cnrm.googlesource.com/cnrm/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:232 +0xae\n\tcnrm.googlesource.com/cnrm/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker(0xc00263e600)\n\t\t/go/src/cnrm.googlesource.com/cnrm/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:211 +0x2b\n\tcnrm.googlesource.com/cnrm/vendor/k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc002b6cad0)\n\t\t/go/src/cnrm.googlesource.com/cnrm/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x5f\n\tcnrm.googlesource.com/cnrm/vendor/k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc002b6cad0, 0x41204c0, 0xc002b771d0, 0xc002599501, 0xc0006e5e00)\n\t\t/go/src/cnrm.googlesource.com/cnrm/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0xa3\n\tcnrm.googlesource.com/cnrm/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc002b6cad0, 0x3b9aca00, 0x0, 0x1, 0xc0006e5e00)\n\t\t/go/src/cnrm.googlesource.com/cnrm/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0xe2\n\tcnrm.googlesource.com/cnrm/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc002b6cad0, 0x3b9aca00, 0xc0006e5e00)\n\t\t/go/src/cnrm.googlesource.com/cnrm/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90 +0x4d\n\tcreated by cnrm.googlesource.com/cnrm/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\t\t/go/src/cnrm.googlesource.com/cnrm/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:193 +0x305\n\t"}
```

**Comment by toumorokoshi** [comment]
Hi! We have a fix that has rolled out in version 1.33.0, please try it out.

If you can reproduce the issue, in that version, please re-open this issue. Thanks!

**Comment by dinvlad** [comment]
Any word on when GKE will be updated to this version? Our is on Regular channel, which still uses `1.31.1`. Thanks!
